### PR TITLE
AC-26: Support for Payara Micro Remote exeuction

### PR DESCRIPTION
**Pull request by pdudits**
Adds profile payara-micro-remote that runs single Payara Micro instance for a test suite.

Profile payara-micro-remote-control serves for starting and stopping such instance.

Currently few tests fail: Security, because there's no security management in payara-micro-deployer, and EJB timers, that might hint at bug in deployment process

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -48,7 +48,7 @@
             (these are downloaded and installed in these versions by Maven for the CI profiles)
          -->
         <payara.version>4.1.2.181</payara.version>
-        <payara.micro.version>5.182</payara.micro.version>
```
Comment (MarkWareham) [inline pom.xml]: Note that we plan to replace this property with just payara.version as the two are never needed at the same time

File: pom.xml [hunk 2]
```diff
@@ -657,6 +657,116 @@
...
+                <!-- default port -->
+                <payara.httpPort>8080</payara.httpPort>
+                <!-- JMS and JAXWS not supported by Micro -->
+                <skipJMS>true</skipJMS>
+                <skipJAXWS>true</skipJAXWS>
+
+                <!-- Being a WebProfile++, Micro does not support ear archives -->
+                <skipEAR>true</skipEAR>
+
+                <!-- Client-cert needs complicated cert setup, which hasn't been done -->
```
Comment (MarkWareham) [inline pom.xml]: "which hasn't been done". Is that out of scope, or a planned future change?

File: pom.xml [hunk 3]
```diff
@@ -657,6 +657,116 @@
...
+                <dependency>
+                    <groupId>fish.payara.arquillian</groupId>
+                    <artifactId>payara-client-ee7</artifactId>
+                </dependency>
+
+                <!-- The actual Arquillian connector -->
+                <dependency>
+                    <groupId>fish.payara.arquillian</groupId>
+                    <artifactId>payara-micro-remote</artifactId>
+                    <version>1.0.Beta4-SNAPSHOT</version>
```
Comment (MarkWareham) [inline pom.xml]: should we be using a snapshot?

File: pom.xml [hunk 4]
```diff
@@ -657,6 +657,116 @@
...
+                                <skipEAR>${skipEAR}</skipEAR>
+                            </systemPropertyVariables>
+                        </configuration>
+                    </plugin>
+                </plugins>
+            </build>
+        </profile>
+
+        <profile>
+            <!-- Helper profile for launching the server. Before the tests run
```
Comment (MarkWareham) [inline pom.xml]: Little confused by this. You are using maven to stop start an instance. Isn't that pretty much just managed?

File: pom.xml (hunk 2 above)
Comment (pdudits) [reply pom.xml]: Mostly, it's a copy-paste from managed profile. But also I wouldn't consider it in scope if managed profile cannot do it.

File: pom.xml (hunk 3 above)
Comment (pdudits) [reply pom.xml]: No we shouldn't, but this serves as a test case that payara/ecosystem-arquillian-connectors#1 works.

I'll file an issue for releasing 1.1 and then would update it.

Comment (Pandrex247) [reply pom.xml]: Properties for versions are nice :) Everything else uses them

File: ejb/remote/roles-allowed/pom.xml [hunk 5]
```diff
@@ -45,6 +45,20 @@
                 </dependency>
             </dependencies>
         </profile>
+
+        <profile>
+            <id>payara-micro-remote</id>
```
Comment (Pandrex247) [inline ejb/remote/roles-allowed/pom.xml]: Other cases of skipping are done using a `skipX` property, is there a reason why you haven't used this pattern here?

File: pom.xml (hunk 4 above)
Comment (Pandrex247) [reply pom.xml]: Is this by-chance optional? Could I still deploy Payara Micro and deploy the micro-deployer manually? I can see why you've made the profile, but I'm not going to remember those commands :P I personally expect a bit of setup when using remote profiles.

File: pom.xml [hunk 6]
```diff
@@ -657,6 +657,116 @@
...
+                <plugins>
+                    <plugin>
+                        <groupId>org.apache.maven.plugins</groupId>
+                        <artifactId>maven-dependency-plugin</artifactId>
+                        <configuration>
+                            <artifactItems>
+                                <artifactItem>
+                                    <groupId>fish.payara.arquillian</groupId>
+                                    <artifactId>payara-micro-deployer</artifactId>
+                                    <version>1.0.Beta4-SNAPSHOT</version>
```
Comment (Pandrex247) [inline pom.xml]: Would prefer a property

File: pom.xml [hunk 7]
```diff
@@ -657,6 +657,116 @@
...
+                                    <type>war</type>
+                                </artifactItem>
+                            </artifactItems>
+                            <stripVersion>true</stripVersion>
+                        </configuration>
+                    </plugin>
+                    <plugin>
+                        <groupId>fish.payara.maven.plugins</groupId>
+                        <artifactId>payara-micro-maven-plugin</artifactId>
+                        <version>1.0.5</version>
```
Comment (Pandrex247) [inline pom.xml]: See other comments.

File: pom.xml (hunk 4 above)
Comment (pdudits) [reply pom.xml]: Yes, this is optional. You can either remember this one, or launch it e.g. with deployment from maven repo.

File: pom.xml [hunk 8]
```diff
@@ -657,6 +657,116 @@
             </build>
         </profile>
 
+        <profile>
+            <id>payara-micro-remote</id>
+
+            <properties>
+                <!-- default port -->
+                <payara.httpPort>8080</payara.httpPort>
```
Comment (jGauravGupta) [inline pom.xml]: Does it support parallel test execution ?

File: pom.xml (hunk 3 above)
Comment (pdudits) [reply pom.xml]: > Properties for versions are nice :) Everything else uses them

I agree and was about to change it. However profile-dependent dependencies in javaee samples projects seem to be an exception. With `payara-client-ee7` being exception of exception :)

I believe this makes sense from upstream project's point of view -- vendor dependencies should not be in top level dependency management, rather in respective profiles. And making profile-specific dependency management section doesn't bring anything.

But it is possible that every `payara-*` profile would include `payara.arquillian.version` and used it for its ecosystem dependency. Would that be a viable option?

File: pom.xml (hunk 8 above)
Comment (pdudits) [reply pom.xml]: Well since deployments can be done in parallel, I don't see why it wouldn't support parallel execution.

Comment (pdudits) [reply pom.xml]: At least that's as far as theory goes.
Inpractice,  `mvn -T2` fails with classloading errors, just like `-Dparallel=classes` does in projects with multiple test classes. So I must correct the answer to "no, it doesn't".

`-Dparallel=methods` work, though (but may not play well with arquillians `@InOrder` annotation)

Do other profiles support parallel executions?

File: pom.xml (hunk 1 above)
Comment (pdudits) [reply pom.xml]: Adjusted to `payara.version`

File: pom.xml (hunk 2 above)
Comment (pdudits) [reply pom.xml]: Improved the wording

File: ejb/remote/roles-allowed/pom.xml (hunk 5 above)
Comment (pdudits) [reply ejb/remote/roles-allowed/pom.xml]: created skipRemoteEJB property

**Pull request merged**